docs/bos/tutorial/queryapi-ws.md

```diff
@@ -0,0 +1,344 @@
+      action.operations
+        .map((operation) => operation["FunctionCall"])
+        .filter((operation) => operation?.methodName === "set")
+        .map((functionCallOperation) => ({
+          ...functionCallOperation,
+          args: base64decode(functionCallOperation.args),
+          receiptId: action.receiptId, // providing receiptId as we need it
+        }))
+        .filter((functionCall) => {
+          const accountId = Object.keys(functionCall.args.data)[0];
+          return Object.keys(functionCall.args.data[accountId]).includes(
+            "widget"
+          );
+        })
+    );
+
+  if (nearSocialWidgetTxs.length > 0) {
+    console.log("Found BOS Widget Development Activity...");
+    const blockHeight = block.blockHeight;
+    const blockTimestamp = block.header().timestampNanosec;
+    console.log(nearSocialWidgetTxs);
+    await Promise.all(
+      nearSocialWidgetTxs.map(async (widgetEditTx) => {
+        const accountId = Object.keys(widgetEditTx.args.data)[0];
+        const widgetName = Object.keys(
+          widgetEditTx.args.data[accountId]["widget"]
+        )[0];
+
+        console.log(`ACCOUNT_ID: ${accountId}`);
+        console.log(widgetName);
+        await handleWidgetTx(
+          accountId,
+          widgetName,
+          blockHeight,
+          blockTimestamp,
+          widgetEditTx.receiptId
+        );
+        console.log(widgetEditTx);
+      })
+    );
+  }
+}
+```
+
+---
+
+This is the JS function that calls the GraphQL mutation `InsertWidgetActivity` and adds a record to the `widget_activity` table:
+
+:::tip
+
+Learn more about [QueryAPI indexing functions](../queryapi/index-function.md) and how to build your own indexers.
+
+:::
+
+```js title=indexerLogic.js
+  async function handleWidgetTx(
+    accountId,
+    widgetName,
+    blockHeight,
+    blockTimestamp,
+    receiptId
+  ) {
+    console.log(accountId, blockHeight, blockTimestamp, receiptId);
+    try {
+      const mutationData = {
+        activity: {
+          account_id: accountId,
+          widget_name: widgetName,
+          block_height: blockHeight,
+          block_timestamp: blockTimestamp,
+          receipt_id: receiptId,
+        },
+      };
+      await context.graphql(
+        `mutation InsertWidgetActivity($activity: roshaan_near_widget_activity_feed_widget_activity_insert_input = {}) {
+  insert_roshaan_near_widget_activity_feed_widget_activity_one(object: $activity) {
+    id
+  }
+}`,
+        mutationData
+      );
+    } catch (e) {
+      console.log(`Could not add widget activity to DB, ${e}`);
+    }
+  }
+```
+
+## Using WebSockets
+
+Once you have a QueryAPI indexer running, you can use WebSockets to get the data in your B.O.S. component. You only need to create a `WebSocket` object pointing to the QueryAPI's GraphQL endpoint.
+
+### Setup
+
+Here's a code snippet from the BOS component that subscribes and processes any activity from the [Widget Activity indexer](#queryapi-indexer):
```

Comment:
I tried to copy and paste the code into the editor to find out that it does not display anything. 

Can we create a note saying that this is just a snippet. and if you would like to play around with the widget, you can see it at this link and fork it?

Reply:
Note added 👍